### PR TITLE
feat: add pre-commit hook to validate catalogValidationValues.json.template file

### PR DESCRIPTION
**Pull request by akocbek**
### Description

Pre-commit hook to validate `catalogValidationValues.json.template` file. Validating:
- is correct JSON file
- are catalog keys part of terraform inputs of the same directory
 
image

### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

### CODE REVIEW

File: module-assets/ci/validateCatalogTemplate.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                f"Catalog key '{catalog_key}' in '{temp_catalog_file}' is not part of \"{root}\" tf inputs."
+            )
+
+    os.remove(
+        os.path.join(root, temp_tf_inputs_json_file)
+    )  # remove temp temp_tf_inputs.json file
+
+
+# find all 'catalogValidationValues.json.template' files in 'solutions' folder
+for root, dirs, files in os.walk("solutions"):
```
Comment (ocofaigh) [inline module-assets/ci/validateCatalogTemplate.py:77]: `catalogValidationValues.json.template` can exist in other folders too. Its not consistent, nor will it ever be as its not mandated to place the code in any specific place. Your going to have to check the full repo

File: module-assets/ci/validateCatalogTemplate.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
+import json
+import os
+import sys
+from subprocess import PIPE, Popen
+
+validation_errors = []
+
+
+# create a new temp file from catalogValidationValues.json.template and replace "$_strings_"
+def create_temp_catalog(root, file):
```
Comment (ocofaigh) [inline module-assets/ci/validateCatalogTemplate.py:10]: create_temp_catalog -> create_temp_json

File: module-assets/.pre-commit-config.yaml [hunk 3]
```diff
@@ -153,6 +153,11 @@ repos:
       entry: python3 ci/terraformDocOverview.py
       language: python
       files: "README.md"
+    - id: validate_catalog
```
Comment (ocofaigh) [inline module-assets/.pre-commit-config.yaml:156]: json_template_lint

File: module-assets/ci/validateCatalogTemplate.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        if file.endswith("catalogValidationValues.json.template"):
+            if ".terraform" not in os.path.join(root, file):
+
+                # create a new temp catalogValidationValues.json.template file
+                temp_catalog_file = create_temp_catalog(root, file)
+
+                # if catalogValidationValues.json.template is not valid JSON file then save the error
+                if not is_json(temp_catalog_file):
+                    validation_errors.append(
+                        os.path.join(root, file) + " is not valid JSON file."
```
Comment (ocofaigh) [inline module-assets/ci/validateCatalogTemplate.py:88]: is not valid JSON file -> is not valid JSON format

File: module-assets/.pre-commit-config.yaml [hunk 5]
```diff
@@ -153,6 +153,11 @@ repos:
...
+      name: Validate catalogValidationValues.json.template file
+      description: Validate catalogValidationValues.json.template file
+      entry: python3 ci/validateCatalogTemplate.py
```
Comment (ocofaigh) [inline module-assets/.pre-commit-config.yaml:159]: validateCatalogTemplate.py -> validateJsonTemplate.py

File: module-assets/.terraform-docs-config-validate-catalog.yaml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+formatter: "json"
```
Comment (ocofaigh) [inline module-assets/.terraform-docs-config-validate-catalog.yaml:1]: can you rename the file `module-assets/.terraform-docs-config-validate-json-template.yaml`

File: module-assets/ci/validateCatalogTemplate.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    with open(os.path.join(root, temp_tf_inputs_json_file)) as json_data:
+        data = json.load(json_data)
+        for tf_input in data["inputs"]:
+            tf_inputs_name.append(tf_input["name"])
+
+    # check if catalog key is part of terraform input variables of the same directory
+    for catalog_key in catalog_keys:
+        if catalog_key not in tf_inputs_name:
+            validation_errors.append(
+                f"Catalog key '{catalog_key}' in '{temp_catalog_file}' is not part of \"{root}\" tf inputs."
```
Comment (ocofaigh) [inline module-assets/ci/validateCatalogTemplate.py:68]: Its not a catalog key, its an input variable. Perhaps just say `"Key '{catalog_key}' in ...."`

**Review by ocofaigh — CHANGES_REQUESTED** [review]
wow very powerful linter, I like it - a few comments

**Comment by akocbek** [comment]
searching for `catalogValidationValues.json.template` in all dirs

image

File: module-assets/ci/validateCatalogTemplate.py (hunk 1 above)
Comment (akocbek) [reply module-assets/ci/validateCatalogTemplate.py]: fixed

File: module-assets/ci/validateCatalogTemplate.py (hunk 2 above)
Comment (akocbek) [reply module-assets/ci/validateCatalogTemplate.py]: fixed

File: module-assets/.pre-commit-config.yaml (hunk 3 above)
Comment (akocbek) [reply module-assets/.pre-commit-config.yaml]: fixed

File: module-assets/ci/validateCatalogTemplate.py (hunk 4 above)
Comment (akocbek) [reply module-assets/ci/validateCatalogTemplate.py]: fixed

File: module-assets/.pre-commit-config.yaml (hunk 5 above)
Comment (akocbek) [reply module-assets/.pre-commit-config.yaml]: fixed

File: module-assets/.terraform-docs-config-validate-catalog.yaml (hunk 6 above)
Comment (akocbek) [reply module-assets/.terraform-docs-config-validate-catalog.yaml]: fixed

File: module-assets/ci/validateCatalogTemplate.py (hunk 7 above)
Comment (akocbek) [reply module-assets/ci/validateCatalogTemplate.py]: fixed

File: module-assets/ci/validateJsonTemplate.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    with open(os.path.join(root, temp_tf_inputs_json_file)) as json_data:
+        data = json.load(json_data)
+        for tf_input in data["inputs"]:
+            tf_inputs_name.append(tf_input["name"])
+
+    # check if catalog_template key is part of terraform input variables of the same directory
+    for catalog_template_key in catalog_template_keys:
+        if catalog_template_key not in tf_inputs_name:
+            validation_errors.append(
+                f"Key '{catalog_template_key}' in '{temp_catalog_template_file}' is not part of \"{root}\" tf inputs."
```
Comment (ocofaigh) [inline module-assets/ci/validateJsonTemplate.py:68]: ok looking at the sceenshot you added, maybe lets reword so it displays like:

'var1' defined in 'solutions/standard/catalogValidationValues.json.template' is not a valid input variable for the terraform in this directory.

Also please be aware that you are currently outputting the the value of `temp_catalog_template_file` which is actually the temp file name, so I don't suggest you include that in the output

```suggestion
                f"'{catalog_template_key}' defined in '{temp_catalog_template_file}' is not part of \"{root}\" tf inputs."
```

File: module-assets/ci/validateJsonTemplate.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    with open(os.path.join(root, temp_tf_inputs_json_file)) as json_data:
+        data = json.load(json_data)
+        for tf_input in data["inputs"]:
+            tf_inputs_name.append(tf_input["name"])
+
+    # check if catalog_template key is part of terraform input variables of the same directory
+    for catalog_template_key in catalog_template_keys:
+        if catalog_template_key not in tf_inputs_name:
+            validation_errors.append(
+                f"Key '{catalog_template_key}' defined in '{temp_catalog_template_file}' is not part of \"{root}\" tf inputs."
```
Comment (ocofaigh) [inline module-assets/ci/validateJsonTemplate.py:68]: Your still printing `temp_catalog_template_file` which is going to be called `temp_catalogValidationValues.json.template` which may cause confusion.

Comment (akocbek) [reply module-assets/ci/validateJsonTemplate.py]: oh I missed that one, fixed now

**Comment by ocofaigh** [comment]
@akocbek Looks good, got a screenshot of final output?

**Comment by akocbek** [comment]
the latest output when hook fails
image

**Comment by ocofaigh** [comment]
@akocbek The output in screenshot looks odd, specifically this line: 
`./solutions/standard copy/catalogValidationValues.json.template is not valid JSON format`

Should it not read: `./solutions/standard/catalogValidationValues.json.template is not valid JSON format`

**Review by ocofaigh — APPROVED** [review]

**Pull request merged**